OrbitCore/LinuxPerfRingBuffer.cpp

```diff
@@ -0,0 +1,145 @@
+#include "PrintVar.h"
+#include "LinuxPerfRingBuffer.h"
+#include "LinuxPerfUtils.h"
+
+#include <assert.h>
+#include <linux/perf_event.h>
+#include <sys/mman.h>
+
+//-----------------------------------------------------------------------------
+LinuxPerfRingBuffer::LinuxPerfRingBuffer(uint32_t a_PerfFileDescriptor)
+{
+    void* mmap_ret = mmap_mapping(a_PerfFileDescriptor);
+
+    // the memory directly before the ring buffer contains the metadata
+    m_Metadata = reinterpret_cast<perf_event_mmap_page*>(mmap_ret);
+    m_BufferLength = m_Metadata->data_size;
+
+    // the buffer length must be a power of 2, so find the exponent in order 
+    // to later optimize the code using shifts/masks instead of divs/mods.
+    uint32_t exponent = 0;
+    uint64_t rest = m_BufferLength;
+    while( rest >>= 1) ++exponent;
+    m_BufferLengthExponent = exponent;
+
+    // beginning of the ring buffer
+    m_Buffer = reinterpret_cast<char*>(mmap_ret) + m_Metadata->data_offset; 
+}
+
+//-----------------------------------------------------------------------------
+LinuxPerfRingBuffer::LinuxPerfRingBuffer(LinuxPerfRingBuffer&& other) noexcept
+{
+    m_Metadata = other.m_Metadata;
+    m_Buffer = other.m_Buffer;
+    m_BufferLength = other.m_BufferLength;
+    m_BufferLengthExponent = other.m_BufferLengthExponent;
+    other.m_Metadata = nullptr;
+    other.m_Buffer = nullptr;
+    other.m_BufferLength = 0;
+    other.m_BufferLengthExponent = 0;
+}
+
+//-----------------------------------------------------------------------------
+LinuxPerfRingBuffer& LinuxPerfRingBuffer::operator=(LinuxPerfRingBuffer&& other) noexcept
+{
+    std::swap(m_Metadata, other.m_Metadata);
```

Comment:
The swap is a bit of a surprise. Why not move?

Reply:
We do not move, as we need to keep the other buffer intact, until the destructor calls munmap.